File: 1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.py
class Solution:
    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        def solve(node, is_root):
            if not node: return None
            if is_root and node.val not in targets:
                result.append(node)
            node.left = solve(node.left, node.val in targets)
            node.right = solve(node.right, node.val in targets)
            return None if node.val in targets else node


        result = list()
        targets = set(to_delete)
        solve(root, True)
        return result

File: 1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.java
class Solution {
    private List<TreeNode> result;
    private Set<Integer> targets;

    private TreeNode solve(TreeNode node, boolean is_root) {
        if (node == null) return null;
        if (is_root && !targets.contains(node.val)) result.add(node);
        node.left = solve(node.left, targets.contains(node.val));
        node.right = solve(node.right, targets.contains(node.val));
        return targets.contains(node.val) ? null : node;
    }

    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        result = new ArrayList<>();
        targets = new HashSet<>();

        for (int target: to_delete) {
            targets.add(target);
        }

        solve(root, true);
        return result;
    }
}